include/secp256k1_musig.h

```diff
@@ -0,0 +1,417 @@
+ *
+ * Observe that an independent counter is needed for each concurrent signing session
+ * such a device is involved in. To avoid fragility, it is therefore recommended that
+ * any offline signer be usable for only a single session at once.
+ *
+ * Given access to such a counter, its output should be used as (or mixed into) the
+ * session ID to ensure uniqueness.
+ *
+ * Fields:
+ *      combined_pk: MuSig-computed combined public key
+ *        n_signers: Number of signers
+ *          pk_hash: The 32-byte hash of the original public keys
+ *   combined_nonce: Summed combined public nonce (undefined if `nonce_is_set` is false)
+ *     nonce_is_set: Whether the above nonce has been set
+ * nonce_is_negated: If `nonce_is_set`, whether the above nonce was negated after
+ *                   summing the participants' nonces. Needed to ensure the nonce's y
+ *                   coordinate has a quadratic-residue y coordinate
+ *              msg: The 32-byte message (hash) to be signed
+ *       msg_is_set: Whether the above message has been set
+ *  has_secret_data: Whether this session object has a signers' secret data; if this
+ *                   is `false`, it may still be used for verification purposes.
+ *           seckey: If `has_secret_data`, the signer's secret key
+ *         secnonce: If `has_secret_data`, the signer's secret nonce
+ *            nonce: If `has_secret_data`, the signer's public nonce
+ *        nonce_commitments_hash: If `has_secret_data` and
+ *                   `nonce_commitments_hash_is_set`, the hash of all signers'
+ *                   commitments
+ * nonce_commitments_hash_is_set: If `has_secret_data`, whether the
+ *                   nonce_commitments_hash has been set
+ */
+typedef struct {
+    secp256k1_pubkey combined_pk;
+    size_t n_signers;
+    unsigned char pk_hash[32];
+    secp256k1_pubkey combined_nonce;
+    int nonce_is_set;
+    int nonce_is_negated;
+    unsigned char msg[32];
+    int msg_is_set;
+    int has_secret_data;
+    unsigned char seckey[32];
+    unsigned char secnonce[32];
+    secp256k1_pubkey nonce;
+    unsigned char nonce_commitments_hash[32];
+    int nonce_commitments_hash_is_set;
+} secp256k1_musig_session;
+
+/** Data structure containing data on all signers in a session.
+ *
+ * The workflow for this structure is as follows:
+ *
+ * 1. This structure is initialized with `secp256k1_musig_pubkey_combine`, which sets
+ *    the `pubkey` and `index` fields, and zeros out all other fields.
+ *
+ * 2. If the signer data is used in a non-public session, it should be updated with
+ *    `secp256k1_musig_get_public_nonce` which in addition to returning the nonce
+ *    takes an array of signer data structs and an array of commitments, and sets
+ *    each signer data struct's `nonce_commitment` field to the corresponding
+ *    commitment. If the signer data is used in a public session, the
+ *    `nonce_commitment` is set set in musig_session_initialize_public.
```

Comment:
I think this comment should be rearranged so that it says, in order,

1. In a public session, use `musig_session_initialize_public` to initialize the `nonce_commitment` field.
2. In a non-public session (signing session? participatory session?) the function `secp256k1_musig_get_public_nonce` should also be used, which as a side-effect also returns the signer's public nonce. This ensures that the public nonce is not exposed until all commitments have been received.

Reply:
I feel like it's a bit weird that we describe a "workflow of a data structure" which kind of explains how to use the module but not entirely. How to use the module isn't explained anywhere else at all, maybe we should do that instead.

Reply:
Agreed, it'd be good to write a `.md` with some instructions. I'm working on a threshold signature scheme that reuses as much of this module as possible, and it's weird to find myself referring back to data structures' comments to remember how everything fits together.

Reply:
I rephrased the comment according to your suggestion.